### PR TITLE
[Typescript] Fix ListContextProvider types

### PR DESCRIPTION
**Pull request by septentrion-730n**

### CODE REVIEW

File: packages/ra-core/src/controller/list/useListController.ts [hunk 1]
```diff
@@ -219,7 +219,7 @@ export interface ListControllerResult<RecordType extends RaRecord = any> {
     onUnselectItems: () => void;
     page: number;
     perPage: number;
-    refetch: UseGetListHookValue<RecordType>['refetch'];
+    refetch: (() => void) | UseGetListHookValue<RecordType>['refetch'];
```
Comment (djhi) [inline packages/ra-core/src/controller/list/useListController.ts:222]: I think we should remove the `(() => void)`

File: packages/ra-core/src/form/choices/ChoicesContext.ts [hunk 2]
```diff
@@ -24,7 +25,7 @@ export type ChoicesContextValue<RecordType extends RaRecord = any> = {
     isLoading: boolean;
     page: number;
     perPage: number;
-    refetch: () => void;
+    refetch: (() => void) | UseGetListHookValue<RecordType>['refetch'];
```
Comment (djhi) [inline packages/ra-core/src/form/choices/ChoicesContext.ts:28]: Why `UseGetListHookValue`?

**Review by djhi — CHANGES_REQUESTED** [review]

File: packages/ra-core/src/controller/list/useListController.ts (hunk 1 above)
Comment (septentrion-730n) [reply packages/ra-core/src/controller/list/useListController.ts:222]: Done to  continue the compatibility between `ListContext` and `ChoicesContext` for `DatagridInput`

File: packages/ra-core/src/form/choices/ChoicesContext.ts (hunk 2 above)
Comment (septentrion-730n) [reply packages/ra-core/src/form/choices/ChoicesContext.ts:28]: Idem -> Done to  continue the compatibility between `ListContext` and `ChoicesContext` for `DatagridInput`

**Review by djhi — APPROVED** [review]

**Pull request merged**